crates/sparrow-batch/src/batch.rs

```diff
@@ -26,6 +32,50 @@ impl Batch {
         }
     }
 
+    /// Construct a new batch, inferring the upper bound from the data.
+    ///
+    /// It is expected that that data is sorted.
+    pub fn try_new_from_batch(data: RecordBatch) -> error_stack::Result<Self, Error> {
+        error_stack::ensure!(
+            data.num_rows() > 0,
+            Error::internal_msg("Unable to create batch from empty data".to_owned())
+        );
+
+        let time_column: &TimestampNanosecondArray = data.column(0).as_primitive();
+        let up_to_time = RowTime::from_timestamp_ns(time_column.value(time_column.len() - 1));
+
+        #[cfg(debug_assertions)]
+        validate(&data, up_to_time)?;
+
+        let time: &TimestampNanosecondArray = data.column(0).as_primitive();
+        let min_present_time = time.value(0);
+        let max_present_time = time.value(time.len() - 1);
+
+        let time = data.column(0).clone();
+        let subsort = data.column(1).clone();
+        let key_hash = data.column(2).clone();
+
+        let schema = data.schema();
+        let fields: Vec<Arc<Field>> = schema.fields()[3..].to_vec();
+        let fields = Fields::from(fields);
```

Comment:
The fields themselves have an `Arc`, so it would probably be better if we pass in the fields (rather than cloning each field separately and making a separate `Fields`). Can probably defer though.

Reply:
Hm..I updated to
```
        let fields = &schema.fields()[3..];
```
but I don't think that that is what you meant. I suppose it avoids allocating for the `vec` at least?

Reply:
TODO: add to doc, file issue. 

Issues for acceptance